sched/wdog/wd_start.c

```diff
@@ -204,15 +214,31 @@ int wd_start(FAR struct wdog_s *wdog, sclock_t delay,
   up_getpicbase(&wdog->picbase);
   wdog->arg = arg;
 
-  /* Calculate delay+1, forcing the delay into a range that we can handle */
+  /* Delay for at least one tick should be performed */
 
-  if (delay <= 0)
+  if (delay == 0)
     {
       delay = 1;
     }
-  else if (++delay <= 0)
+  else
     {
-      delay--;
+      /* Calculate delay+1, forcing the delay into a range that we can
+       * handle.  In general wd_start() can be called at any time and we do
+       * not know is it at the beginning in the middle or at the end of a
+       * system clock phase, so we need to add 1 tick to ensure that the
+       * operation will be delayed for at least delay number of ticks.  But
+       * in a special case when wd_start() is called from a wd expiration
+       * callback that is called from a system tick interrupt handler we know
+       * that we are exactly at the beginning of a new phase so we do not
+       * need to add an extra tick in this case.
+       */
+
+      delay += g_wdexpiration_context ? 0 : 1;
```

Comment:
This approach has two issues:

1. If some wdog callback run a bit long, the phase error will be large
2. Can't fix all possible place(e.g. other interrupt happen near the timer interrupt)

Since the accuracy of wdog is one tick, the change can't improve the accuracy smaller than one tick in general from the statistics perspective. But it improve in the most common case, I am fine with the change.

Reply:
Maybe it is good to move this under a feature flag? So the user will enable this behavior explicitly. By default it will be not enabled.

Reply:
@pkarashchenko enabling it be default will bring benefits or drawbacks? I think we need to reduce the number of complex configurations inside NuttX, not create more configurations. The menu configuration "entropy" is becoming higher and higher...

Reply:
Since this modification fixes the common case, it should be enabled by default.

Reply:
> This approach has two issues:
> 
> 1. If some wdog callback run a bit long, the phase error will be large
> 2. Can't fix all possible place(e.g. other interrupt happen near the timer interrupt)
> 
> Since the accuracy of wdog is one tick, the change can't improve the accuracy smaller than one tick in general from the statistics perspective. But it improve the most common case, I am fine with the change.

I can agree with 1, not with 2. Even if separate HW timer is used the other interrupt can happen near the timer interrupt. 2 is usually mitigated by assigning of a proper interrupt priority and enabling nested interrupts, so can't be a valid argument in this discussion.

Reply:
> Since this modification fixes the common case, it should be enabled by default.

I'm almost convinced to close both of my PRs and start implementation of high resolution timers ;)

Reply:
> 
> I can agree with 1, not with 2. Even if separate HW timer is used the other interrupt can happen near the timer interrupt. 2 is usually mitigated by assigning of a proper interrupt priority and enabling nested interrupts, so can't be a valid argument in this discussion.

Why do you assume other interrupt happen near the timer interrupt? Many interrupts are trigger by the external activity, nobody can forecast when the activity will happen.

Reply:
> > 
> > I can agree with 1, not with 2. Even if separate HW timer is used the other interrupt can happen near the timer interrupt. 2 is usually mitigated by assigning of a proper interrupt priority and enabling nested interrupts, so can't be a valid argument in this discussion.
> 
> Why do you assume other interrupt happen near the timer interrupt? Many interrupts are trigger by the external activity, nobody can forecast when the activity will happen.

I'm assuming that it may happen, but external interrupts may happen near other MCU timer interrupt. The usage if a dedicated HW timer was suggested as a solution to have more accurate timers. What I'm trying to say that in terms of 2 there is no difference with sys tick interrupt